client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -84,30 +120,36 @@ private final class PoolManager[F[_], A <: Connection[F]](
         if (!isClosed) {
           @tailrec
           def go(): Unit =
-            idleQueue.dequeueFirst(_.requestKey == key) match {
+            getConnectionFromQueue(key) match {
               case Some(conn) if !conn.isClosed =>
                 logger.debug(s"Recycling connection: $stats")
                 callback(Right(NextConnection(conn, fresh = false)))
 
               case Some(closedConn) =>
                 logger.debug(s"Evicting closed connection: $stats")
-                allocated -= 1
+                decrConnection(key)
                 go()
 
-              case None if allocated < maxTotal =>
+              case None if numConnectionsCheckHolds(key) =>
                 logger.debug(s"Active connection not found. Creating new one. $stats")
                 createConnection(key, callback)
 
-              case None if idleQueue.nonEmpty =>
+              case None if curTotal == maxTotal =>
                 logger.debug(
                   s"No connections available for the desired key. Evicting oldest and creating a new connection: $stats")
-                allocated -= 1
-                idleQueue.dequeue().shutdown()
-                createConnection(key, callback)
+                val keys = idleQueues.keys
+                if (keys.nonEmpty) {
+                  val randKey = keys.drop(Random.nextInt(keys.size)).head
+                  idleQueues(randKey).dequeue().shutdown()
```

Comment:
As I understand this, we can have an empty `idleQueue` if there is a connection outstanding for its key, and we only remove the queue if the last key belonging to it is removed. That would mean that the `.dequeue()` call could throw a `NoSuchElementException` and we should probably go with `getConnectionFromQueue`.

In this case, should we be now randomly looking at a different queue? I feel like this scheme will get pretty ugly without just guaranteeing that there are no empty queues in the `idleQueues` map.

Reply:
Sorry, I don't see how we can have an empty queue in idleQueues 👅 . Either the entire map would be empty or it will only have non empty queues.

Reply:
When we successfully get a live connection with `getConnectionFromQueue`, it removes a connection from the queue but doesn't remove the queue if it's empty.

Reply:
OOPS !!!